### ISSUE TITLE
Export Models to ONNX and Serve Batch GNN Inference via Triton

### ISSUE THREAD

**Issue by Inkman007** [Ledger-Lenz/Ledgerlens-core#317]
## Summary
`detection/model_inference.py` and the GNN modules (`detection/gnn_model.py`, `detection/gnn_ring_detector.py`) run all inference in-process using joblib-loaded scikit-learn-API models and PyTorch, respectively. This is a latency and throughput ceiling for batch scoring workloads such as `api/batch_router.py`, which fans out per-wallet scoring across up to 1,000 wallets per job. Export the ensemble and GNN models to ONNX and add an optional NVIDIA Triton Inference Server backend behind the existing scoring interface, with in-process inference remaining the default for local development.

## Background & Context
Real-time single-wallet scoring in `detection/model_inference.py` loads `random_forest.joblib`, `xgboost.joblib`, and `lightgbm.joblib` via `safe_joblib_load` (see `detection/model_signing.py`) and calls `.predict_proba()` in-process for every request. The GNN path (`detection/gnn_ring_detector.py`, `GNNRingDetector`) similarly loads a PyTorch checkpoint via `safe_load_gnn_checkpoint` and runs `GraphSAGEEncoder` + `RingMembershipClassifier` forward passes directly in the API process. `docs/performance.md` documents scale targets for the graph engine (e.g., 100K nodes / 500K edges in ~27s on a single core) — the same single-process constraint applies to model inference, and there is currently no path to horizontal scale-out for the ML scoring step itself.

`api/batch_router.py` (`BatchRequest`, `_process_batch`) already queues up to 1,000-wallet batch jobs with a concurrency cap (`_MAX_CONCURRENT = 5`) and delegates to `get_latest_scores`, but the actual model forward pass for each wallet is still a Python-process-bound call. At the scale targets described in `docs/performance.md`, this makes Python/GIL-bound in-process inference — not graph construction — the eventual bottleneck for batch jobs, since neither joblib-based sklearn inference nor PyTorch CPU inference release the GIL efficiently under concurrent asyncio scheduling.

This issue is purely about the serving layer, not model architecture: it does not change what the RF/XGBoost/LightGBM ensemble or the GNN encoder/classifier predict, only how their forward pass is executed at scale. It must preserve the model-signing integrity guarantees in `detection/model_signing.py` (`assert_within_model_dir`, `safe_joblib_load`) and the GNN checksum verification in `detection/gnn_ring_detector.py` — ONNX export happens *after* signature/checksum verification of the source artifact, and the exported `.onnx` file gets its own signature so a compromised export pipeline cannot silently substitute a different model.

## Objectives
- [ ] Add an `scripts/export_onnx.py` CLI script that converts each trained `.joblib` model (via `skl2onnx` / `onnxmltools` for RF, and native ONNX export for XGBoost/LightGBM) and the GNN checkpoint (via `torch.onnx.export`) into versioned `.onnx` artifacts under `models/onnx/`
- [ ] Define an `InferenceBackend` protocol with `InProcessBackend` (current behavior, default) and `TritonBackend` (gRPC/HTTP client to a Triton server) implementations, selected via configuration
- [ ] Wire `TritonBackend` into `detection/model_inference.py`'s scoring entry point and `GNNRingDetector.score()` without changing their public return types (`RiskScore` fields, GNN membership score in `[0, 1]`)
- [ ] Sign each exported `.onnx` file using the existing ED25519 `ModelSigner` scheme (`docs/model_signing.md`) so ONNX artifacts inherit the same integrity/authenticity guarantees as `.joblib`/`.pt` files
- [ ] Add a `docker-compose.triton.yml` profile that runs a local Triton container alongside the existing `docker-compose.yml` API service for integration testing
- [ ] Benchmark in-process vs. Triton latency/throughput for a representative batch (1,000 wallets) and record results in `docs/performance.md`
- [ ] Add graceful fallback: if `TritonBackend` fails health checks or a request errors, fall back to `InProcessBackend` and log a `WARNING` (never fail the batch job outright)

## Technical Requirements

**Configuration** (add to `config/settings.py`, following the existing lower_snake_case field / UPPER_SNAKE_CASE env var convention seen with `gnn_model_path` / `GNN_MODEL_PATH`):
```python
inference_backend: str = "in_process"        # "in_process" | "triton"
triton_url: str = "localhost:8001"            # gRPC endpoint
triton_model_repository: str = "models/onnx"
triton_request_timeout_seconds: float = 2.0
triton_health_check_interval_seconds: float = 30.0
onnx_export_dir: str = "models/onnx"
```

**`InferenceBackend` protocol:**
```python
# detection/inference_backend.py
from typing import Protocol
import numpy as np

class InferenceBackend(Protocol):
    def predict_proba(self, model_name: str, features: np.ndarray) -> np.ndarray:
        """features: shape (n_samples, n_features). Returns shape (n_samples, 2)."""
        ...

    def health_check(self) -> bool: ...


class TritonBackend:
    def __init__(self, url: str, timeout_seconds: float = 2.0):
        import tritonclient.grpc as grpcclient
        self._client = grpcclient.InferenceServerClient(url=url)
        self._timeout = timeout_seconds

    def predict_proba(self, model_name: str, features: np.ndarray) -> np.ndarray:
        # Build InferInput/InferRequestedOutput, call self._client.infer(...)
        # Raise InferenceBackendError on timeout or non-200; caller falls back.
        ...

    def health_check(self) -> bool:
        try:
            return self._client.is_server_ready()
        except Exception:
            return False
```

**ONNX export sketch** (`scripts/export_onnx.py`):
```python
from skl2onnx import to_onnx
from detection.model_signing import safe_joblib_load, sign_model_file

def export_sklearn_model(model_path: Path, output_path: Path, n_features: int) -> None:
    model = safe_joblib_load(model_path)  # verifies existing .sig first
    onnx_model = to_onnx(model, initial_types=[("input", FloatTensorType([None, n_features]))])
    output_path.write_bytes(onnx_model.SerializeToString())
    sign_model_file(output_path)  # new .onnx.sig, ED25519, same scheme as docs/model_signing.md
```

Selection logic in `model_inference.py` reads `settings.inference_backend`; `TritonBackend` instances perform `health_check()` at startup and every `TRITON_HEALTH_CHECK_INTERVAL_SECONDS`, downgrading to `InProcessBackend` automatically on failure (mirrors the `SorobanPublisher` circuit-breaker pattern already used for on-chain submission).

## Security Considerations
- Triton's gRPC/HTTP endpoints must never be exposed outside the internal network; `TRITON_URL` should be validated to reject public/non-loopback hosts unless `TRITON_ALLOW_REMOTE=true` is explicitly set, mirroring the `https://`-only / SSRF-style validation already applied to `EVM_PROVIDERS` in `config/settings.py`
- ONNX export must run on the **verified** in-memory model object returned by `safe_joblib_load`/`safe_load_gnn_checkpoint` (post signature/checksum check), never re-read the raw `.joblib`/`.pt` bytes independently — this closes a gap where a tampered source file could still produce a "validly" exported ONNX model
- Exported `.onnx` files are signed with the same ED25519 `ModelSigner` key material described in `docs/model_signing.md`; `verify-models` (`cli.py`) must be extended to also verify `.onnx.sig` files
- No raw feature vectors, wallet addresses, or scores are logged in Triton client request/response paths — wallet addresses passed to `predict_proba` call sites remain masked per the existing logging convention (`GABC1234...WXYZ`)
- Fallback-to-in-process must be a fail-open decision only for the backend choice, not for model integrity — a Triton response must still be validated for shape/range before being treated as trusted (`score ∈ [0,1]`, correct output tensor shape) to avoid a compromised or misconfigured Triton deployment silently corrupting scores

## Testing Requirements
- `tests/test_inference_backend.py`: `InProcessBackend` behavior is unchanged (regression against current `model_inference.py` output for a fixed feature vector)
- `tests/test_inference_backend.py`: `TritonBackend.predict_proba` against a mocked gRPC client returns the expected shape and raises `InferenceBackendError` on timeout
- `tests/test_inference_backend.py`: automatic fallback from `TritonBackend` to `InProcessBackend` on `health_check()` failure, with a `WARNING` log emitted
- `tests/test_onnx_export.py`: exported RF/XGBoost/LightGBM ONNX models produce predictions within `1e-4` absolute tolerance of the source `.joblib` model on a fixed synthetic feature matrix
- `tests/test_onnx_export.py`: `export_sklearn_model` refuses to export when the source `.joblib` fails signature verification (raises `ModelIntegrityError`)
- `tests/test_model_signing.py` (extend existing suite): `.onnx.sig` files are generated and verified through the same `ModelSigner`/`verify-models` path
- Integration test (marked `slow`): full round-trip against a live Triton container from `docker-compose.triton.yml`, scoring a batch of 1,000 synthetic wallets and comparing throughput against the in-process path

## Documentation Requirements
- `docs/performance.md`: add an "Inference Serving" section with in-process vs. Triton benchmark numbers and when to switch backends
- `docs/model_signing.md`: extend to cover `.onnx.sig` generation/verification
- `README.md`: add a short "Batch Inference Scaling" note under the Machine Learning Layer section pointing to the new docs, and list `INFERENCE_BACKEND` / `TRITON_URL` in the configuration reference
- `.env.example`: add all new `TRITON_*` / `ONNX_EXPORT_DIR` settings with inline comments

## Definition of Done
- [ ] All objectives completed
- [ ] Tests pass (`pytest`)
- [ ] No regressions on existing test suite
- [ ] PR reviewed and approved

## For Contributors
**When applying for this issue, please specify:**
- Your area of specialty
- Relevant experience with: ONNX export tooling (`skl2onnx`, `onnxmltools`, `torch.onnx`), NVIDIA Triton Inference Server, gRPC clients, and production ML serving at scale
- Your approach or initial thoughts on the implementation
- Estimated time to complete

**Ideal contributor profile:** ML infrastructure engineer with hands-on experience deploying and operating Triton (or an equivalent model server) in production, comfortable with both scikit-learn-API and PyTorch export paths, and mindful of keeping the default local-dev path (in-process) unaffected.

**Comment by Inkman007** [comment]
Closing as part of a full issue-tracker reset: the project's open issues have been replaced with a new curated set of 125 advanced-level issues covering the codebase end to end. No action needed here; if this item is still relevant, please open a fresh issue referencing it.
